pkg/routeagent/controllers/route/route.go

```diff
@@ -165,6 +258,153 @@ func (r *Controller) processNextCluster() bool {
+	ipAddr, vtepMask, err := r.getHostIfaceIPAddress()
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve the IPv4 address on the Host %v", err)
+	}
+
+	vtepIP, err := r.getVxlanVtepIPAddress(ipAddr.String())
+	if err != nil {
+		return fmt.Errorf("Failed to derive the vxlan vtepIP %v", err)
+	}
+
+	if ifaceType == VxInterfaceGateway {
+		attrs := &vxLanAttributes{
+			name:     VxLANIface,
+			vxlanId:  100,
+			group:    nil,
+			srcAddr:  nil,
+			vtepPort: VxLANPort,
+			mtu:      1450,
+		}
+
+		r.vxlanDevice, err = newVxlanIface(attrs)
+		if err != nil {
+			return fmt.Errorf("Failed to create vxlan interface on Gateway Node: %v", err)
+		}
+
+		for fdbAddress := range r.remoteVTEPs.Set {
+			err = r.vxlanDevice.AddFDB(net.ParseIP(fdbAddress), "00:00:00:00:00:00")
+			if err != nil {
+				return fmt.Errorf("Failed to add FDB entry on the Gateway Node vxlan iface %v", err)
+			}
+		}
+
+		// Enable loose mode (rp_filter=2) reverse path filtering on the vxlan interface.
+		err = ioutil.WriteFile("/proc/sys/net/ipv4/conf/"+VxLANIface+"/rp_filter", []byte("2"), 0644)
+		if err != nil {
+			return fmt.Errorf("Unable to update vxlan rp_filter proc entry, err: %s", err)
+		} else {
+			klog.Info("Successfully configured rp_filter to loose mode(2) ")
+		}
+
+	} else if ifaceType == VxInterfaceWorker {
+		// non-Gateway/Worker Node
+		attrs := &vxLanAttributes{
+			name:     VxLANIface,
+			vxlanId:  100,
+			group:    gatewayNodeIP,
+			srcAddr:  vtepIP,
+			vtepPort: VxLANPort,
+			mtu:      1450,
+		}
+
+		r.vxlanDevice, err = newVxlanIface(attrs)
+		if err != nil {
+			return fmt.Errorf("Failed to create vxlan interface on non-Gateway Node: %v", err)
+		}
+	}
+
+	err = r.vxlanDevice.configureIPAddress(vtepIP, vtepMask.Mask)
+	if err != nil {
+		return fmt.Errorf("Failed to configure vxlan interface ipaddress on the Gateway Node %v", err)
+	}
+
+	return nil
+}
+
+func (r *Controller) processNextPod() bool {
+	obj, shutdown := r.podWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer r.podWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return fmt.Errorf("Error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		pod, err := r.clientSet.CoreV1().Pods(ns).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("Error retrieving submariner-route-agent pod object %s: %v", name, err)
+		}
+
+		klog.V(4).Infof("In processNextPod, POD HostIP is %s", pod.Status.HostIP)
+		r.populateRemoteVtepIps(pod.Status.PodIP)
+
+		// A new Node (identified via a Submariner-route-agent daemonset pod event) is added to the cluster.
+		// On the GatewayDevice, update the vxlan fdb entry (i.e., remote Vtep) for the newly added node.
+		if r.isGatewayNode {
+			if r.vxlanDevice != nil {
+				err := r.vxlanDevice.AddFDB(net.ParseIP(pod.Status.PodIP), "00:00:00:00:00:00")
```

Comment:
It's possible `r.vxlanDevice` could be nil right after the check on line 383. To avoid this, you should  first store `r.vxlanDevice` in a local var. Or eliminate concurrency as I suggested in another comment.

Reply:
Why will vxlanDevice become nil? Once we create a VxLAN interface on the host, we will not delete it. Am I missing something?

Reply:
In `handleRemovedEndpoint`, you set `r.vxlanDevice` to nil. Let's say thread 1 executes line 394 here and observes  `r.vxlanDevice` non-nil. Thread 2 then interleaves and executes line 566 in `handleRemovedEndpoint` to set it to nil. Thread 1 resumes and executes line 395 but now `r.vxlanDevice` is nil. Doing the following would alleviate that potential issue:

```
localVxlanDevice := r.vxlanDevice
if localVxlanDevice != nil {
    localVxlanDevice. AddFDB(...)
{
```

Reply:
Thanks for the explanation. The chances of endpoint being removed AFAIU is very less (in general). Anyways, lets protect it with a Mutex, to be safe. Please take a look at the updated code.